pkg/executables/kind.go

```diff
@@ -225,35 +254,86 @@ func (k *Kind) setupExecConfig(clusterSpec *cluster.Spec) error {
-		k.execConfig.RegistryMirrorMap = containerd.ToAPIEndpoints(registryMirror.NamespacedRegistryMap)
-		if registryMirror.CACertContent != "" {
-			path := filepath.Join(clusterSpec.Cluster.Name, "generated", "certs.d", registryMirror.BaseRegistry)
-			if err := os.MkdirAll(path, os.ModePerm); err != nil {
-				return err
-			}
-			if err := os.WriteFile(filepath.Join(path, "ca.crt"), []byte(registryMirror.CACertContent), 0o644); err != nil {
-				return errors.New("error writing the registry certification file")
-			}
-			k.execConfig.RegistryCACertPath = filepath.Join(clusterSpec.Cluster.Name, "generated", "certs.d")
-		}
-		if registryMirror.Auth {
-			k.execConfig.RegistryAuth = registryMirror.Auth
-			username, password, err := config.ReadCredentials()
-			if err != nil {
-				return err
-			}
-			k.execConfig.RegistryUsername = username
-			k.execConfig.RegistryPassword = password
+		if err := k.setupRegistryMirror(clusterSpec, registryMirror); err != nil {
+			return err
 		}
 	}
+
 	if err := k.CreateAuditPolicy(clusterSpec); err != nil {
 		return err
 	}
 	return nil
 }
 
+// setupRegistryMirror handles the complete setup of registry mirror configuration.
+func (k *Kind) setupRegistryMirror(clusterSpec *cluster.Spec, registryMirror *registrymirror.RegistryMirror) error {
+	mirrorBase := registryMirror.BaseRegistry
+	registryMirrorMap := containerd.ToAPIEndpoints(registryMirror.NamespacedRegistryMap)
+
+	// Create the base certs.d directory
+	certsBasePath := filepath.Join(clusterSpec.Cluster.Name, "generated", "certs.d")
+	k.execConfig.RegistryConfigDir = certsBasePath
+
+	var username, password string
+	if registryMirror.Auth {
+		var err error
+		username, password, err = config.ReadCredentials()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create CA certificate only once for the base registry and determine shared CA path
+	var mountedCACertPath string
+	if registryMirror.CACertContent != "" {
+		// Create the base registry directory and CA certificate file
+		mirrorBaseDir := filepath.Join(certsBasePath, mirrorBase)
+		if err := os.MkdirAll(mirrorBaseDir, os.ModePerm); err != nil {
+			return fmt.Errorf("error creating directory %s: %w", mirrorBaseDir, err)
+		}
+
+		// Write the CA certificate file
+		caPath := filepath.Join(mirrorBaseDir, "ca.crt")
+		if err := os.WriteFile(caPath, []byte(registryMirror.CACertContent), 0o644); err != nil {
+			return fmt.Errorf("error writing CA certificate: %w", err)
+		}
+
+		mountedCACertPath = fmt.Sprintf("/etc/containerd/certs.d/%s/ca.crt", mirrorBase)
+	}
+
+	// Setup configuration for the mirror registry
+	err := setupRegistryConfig(RegistryConfig{
+		Server:     mirrorBase,
+		Host:       mirrorBase,
+		CACertPath: mountedCACertPath,
+		Username:   username,
+		Password:   password,
+		OutputDir:  filepath.Join(certsBasePath, mirrorBase),
+	})
+	if err != nil {
+		return fmt.Errorf("error setting up configuration for mirror registry %s: %w", mirrorBase, err)
```

Comment:
Can you please specify which registry we are discussing here? Mirror registry meaning the local/private registry?